core/app/models/spree/stock/allocator/base.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module Spree
+  module Stock
+    module Allocator
+      class Base
+        attr_reader :availability
+
+        def initialize(availability)
+          @availability = availability
+        end
+
+        def allocate_inventory(desired)
+          # Allocate any available on hand inventory
+          on_hand, desired = allocate_on_hand(desired)
+
+          # Allocate remaining desired inventory from backorders
+          backordered, desired = allocate_backordered(desired)
+
+          # If all works at this point desired must be empty
+          [on_hand, backordered, desired]
+        end
+
+        private
+
+        def allocate_on_hand
+          raise NotImplementedError
+        end
+
+        def allocate_backorderable
```

Comment:
Can we make this accept an argument, just to show how the interface is supposed to look?